lib/rules/progressbar-needs-labelling.js

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+"use strict";
+
+const { hasFieldParent } = require("../util/hasFieldParent");
+const { hasNonEmptyProp } = require("../util/hasNonEmptyProp");
+const elementType = require("jsx-ast-utils").elementType;
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        // possible error messages for the rule
+        messages: {
+            noUnlabelledProgressbar:
+                "Accessibility: Progressbar must have aria-valuemin, aria-valuemax, aria-valuenow, and aria-describedby attributes"
+        },
+        // "problem" means the rule is identifying code that either will cause an error or may cause a confusing behavior: https://eslint.org/docs/latest/developer-guide/working-with-rules
+        type: "problem",
+        // docs for the rule
+        docs: {
+            description:
+                "Accessibility: Progressbar must have aria-valuemin, aria-valuemax, aria-valuenow, and aria-describedby attributes",
+            recommended: true,
+            url: "https://www.w3.org/TR/html-aria/" // URL to the documentation page for this rule
+        },
+        schema: []
+    },
+    // create (function) returns an object with methods that ESLint calls to “visit” nodes while traversing the abstract syntax tree
+    create(context) {
+        return {
+            // visitor functions for different types of nodes
+            JSXOpeningElement(node) {
+                // if it is not a ProgressBar, return
+                if (elementType(node) !== "ProgressBar") {
+                    return;
+                }
+
+                const mandatoryAttributes = ["aria-valuemin", "aria-valuemax", "aria-valuenow", "aria-describedby"];
+
+                if (mandatoryAttributes.every(attribute => hasNonEmptyProp(node.attributes, attribute)) && hasFieldParent(context)) {
```

Comment:
https://www.magentaa11y.com/checklist-web/progress/
https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/progressbar_role#best_practices

The File component is one of the valid ways to label the progress bar.
If the validationMessage prop is provided then aria-describedby is applied to the ProgressBar, in that case we don't need to check for it.

Other valid ways to label a progress bar include:
- aria-label (although a visual label is preferred)
- aria-labelledby
- does wrapped in Label work?

The min and max values are not required, they will default to 0 and 100 if not specified. 
It might be better to check for the FluentUI props min and max.

Reply:
I see, I will add validation for aria-label, aria-labelledby and remove for min - max props. About using Label as a wrapper - I see in the fluent [implementation](https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-progress/library/src/components/ProgressBar/useProgressBar.tsx) that ProgressBar is using Field component's value and not from Label.

Reply:
@aubreyquinn  if aria-valuemin and aria-valuemax are not defined then it is being set to undefined in the component's code.

Reply:
> I see, I will add validation for aria-label, aria-labelledby and remove for min - max props. About using Label as a wrapper - I see in the fluent [implementation](https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-progress/library/src/components/ProgressBar/useProgressBar.tsx) that ProgressBar is using Field component's value and not from Label.

The recommended way is to use the Field component, but someone might not do that. As we are looking for real WCAG violations then we can check for the other cases too.

Reply:
> @aubreyquinn if aria-valuemin and aria-valuemax are not defined then it is being set to undefined in the component's code.

Strange, I did not observe that in Stackblitz. Maybe there is a bug in the code. In that case, if the React props min/max are missing, we can then also check for those aria attributes.

Reply:
> > I see, I will add validation for aria-label, aria-labelledby and remove for min - max props. About using Label as a wrapper - I see in the fluent [implementation](https://github.com/microsoft/fluentui/blob/master/packages/react-components/react-progress/library/src/components/ProgressBar/useProgressBar.tsx) that ProgressBar is using Field component's value and not from Label.
> 
> The recommended way is to use the Field component, but someone might not do that. As we are looking for real WCAG violations then we can check for the other cases too.

I agree. I have added check for attributes which were being derived from field attribute (like `aria-describedby` and `aria-labelledby`). Please lmk I missed any other attributes.